File: 07-agent/cc-agent/src/agent/session-key.ts
import type { AgentRunParams } from "./model.js";
import { ChannelId } from "../channels/types.js";

export type SessionKeyParts = Pick<AgentRunParams, "channelId" | "appId" | "groupId">;

export function buildSessionKey(parts: SessionKeyParts): string {
  return [
    parts.channelId,
    encodeURIComponent(parts.appId),
    encodeURIComponent(parts.groupId)
  ].join(":");
}

export function parseSessionKey(sessionKey: string): SessionKeyParts {
  const parts = sessionKey.split(":");
  if (parts.length !== 3) {
    throw new Error(`Invalid session_key: ${sessionKey}`);
  }

  const [channelId, encodedAppId, encodedGroupId] = parts;
  if (!isChannelId(channelId) || !encodedAppId || !encodedGroupId) {
    throw new Error(`Invalid session_key: ${sessionKey}`);
  }

  return {
    channelId,
    appId: decodeURIComponent(encodedAppId),
    groupId: decodeURIComponent(encodedGroupId)
  };
}

export function normalizeThreadId(threadId?: string): string {
  return threadId?.trim() || "";
}

export function buildQueueKey(message: SessionKeyParts & { threadId?: string }): string {
  const sessionKey = buildSessionKey(message);
  const threadId = normalizeThreadId(message.threadId);
  return threadId ? `${sessionKey}:${encodeURIComponent(threadId)}` : sessionKey;
}

function isChannelId(value: string): value is ChannelId {
  return value === ChannelId.Feishu || value === ChannelId.Wechat;
}

File: 07-agent/cc-agent/app/agent/session_key.py
from dataclasses import dataclass
from urllib.parse import quote, unquote


SESSION_KEY_PARTS = 3


@dataclass(frozen=True)
class ParsedSessionKey:
    channel_id: str
    app_id: str
    group_id: str


def build_session_key(channel_id: str, app_id: str, group_id: str) -> str:
    return "/".join(quote(part, safe="") for part in (channel_id, app_id, group_id))


def parse_session_key(session_key: str) -> ParsedSessionKey:
    parts = session_key.split("/")
    if len(parts) != SESSION_KEY_PARTS or any(part == "" for part in parts):
        raise ValueError(f"Invalid session_key: {session_key!r}")
    channel_id, app_id, group_id = (unquote(part) for part in parts)
    return ParsedSessionKey(channel_id=channel_id, app_id=app_id, group_id=group_id)

